File: code/Python/ABC231D.py
class UnionFind:
    def __init__(self, n):
        self.par = [-1]*n
        self.siz = [1]*n

    #経路圧縮あり
    def root(self, x):
        if self.par[x] == -1:
            return x

        self.par[x] = self.root(self.par[x])
        return self.par[x]

    #経路圧縮なし
    """     
    def root(self, x):
        if self.par[x] == -1:
            return x
        return self.root(self.par[x])
    """
    
    def unite(self, x, y):
        rx = self.root(x)
        ry = self.root(y)
        if rx == ry:
            return False
        if self.size(rx) < self.size(ry):
            rx, ry = ry, rx

        self.par[ry] = rx
        self.siz[rx] += self.siz[ry]
        return True


    def same(self, x, y):
        return self.root(x) == self.root(y)

    def size(self, x):
        return self.siz[self.root(x)]


n, m = map(int, input().split())
cnt = [0]*n
uf = UnionFind(n)
exist = True

for _ in range(m):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    if uf.unite(a, b) == False:
        exist = False
    cnt[a] += 1
    cnt[b] += 1
    if cnt[a] >= 3 or cnt[b] >= 3:
        exist = False

if exist:
    print("Yes")
else:
    print("No")

exit()

File: code/C++/ABC231D.cpp
#include<iostream>
#include<vector>

class UnionFind {
    public:
        std::vector<int> par, siz;
        UnionFind(int n): par(n, -1), siz(n, 1){}

        //経路圧縮あり
        int root(int x){
            if(par[x] == -1) return x;
            return par[x] = root(par[x]);
        }

        //経路圧縮なし
        /*int root(int x){
            if(par[x] == -1) return x;
            return root(par[x]);
        } */

        bool unite(int x, int y){
            int rx = root(x), ry = root(y);
            if(rx == ry) return false;
            if(size(rx) < size(ry)) std::swap(rx, ry);
            
            par[ry] = rx;
            siz[rx] += siz[ry];
            return true; 
        }

        bool same(int x, int y){
            return root(x) == root(y);
        }

        int size(int x){
            return siz[root(x)];
        }

};

//解答の方針
//隣合う人がひとつに対して3人以上いる、もしくは隣合う関係を辺で結んだ時閉路になると不可能なことがわかります。
//隣合う関係性は有向グラフではなく無向グラフで表せるのでUnion-Findで検出できます。

int main(){
    int n, m;
    std::cin >> n >> m;
    bool exist = true;
    std::vector<int> cnt(n, 0);
    UnionFind uf(n);
    for(int i = 0;i < m;++i){
        int a, b;
        std::cin >> a >> b;
        --a;--b;
        if(uf.unite(a, b) == false) exist = false;
        ++cnt[a];
        ++cnt[b];
        if(cnt[a] >= 3 || cnt[b] >= 3) exist = false;
    }

    if(exist) std::cout << "Yes" << std::endl;
    else std::cout << "No" << std::endl;

    return 0;
}
